### PR TITLE
Add alpha blending support (WebGL and Canvas) to tiles

### PR DESCRIPTION
**Pull request by blurymind**
This adds an alpha option to tiles, which is much needed to improve compatibility with the tiled and ldtk tilemap formats.

image

You can achieve this effect by either passing the option alpha
example: `{alpha: 0.5}`

or by using tileAlpha method
example: 
`tilemap.tile(resources.atlas.textures["chest.png"], 0 * 32, 0).tileAlpha(0.5);`

It also makes the code easier to read  by using Enum instead of hardcoded values for the array offset. This part has always made it kind of confusing to get a value to the actual shader. There is also now an empty line between the offset cycles (arr[sz++]...) which also helps with counting up the index on each cycle.

The PR is a part of an ongoing task to improve compatibility with tiled and also add a new ldtk parser to gdevelop
https://github.com/4ian/GDevelop/issues/2434

### CODE REVIEW

**Comment by blurymind** [comment]
@ivanpopelyshev any idea why the auto build checks are failing?  :)

**Comment by ivanpopelyshev** [comment]
Look in `files changed` tab

**Comment by blurymind** [comment]
Thanks, I sorted out all the warnings that it showed. The first test still fails (Latest PixiJS Node.js CI / build )

**Comment by ivanpopelyshev** [comment]
> Error: src/CanvasTileRenderer.ts(1,39): error TS2307: Cannot find module '@pixi/core' or its corresponding type declarations.

@SukantPal

**Comment by SukantPal** [comment]
Don't look at the pipeline that tests against latest dev. I haven't fully set it up yet. This is passing the first pipeline so it's good.

File: src/TilemapShader.ts [hunk 1]
```diff
@@ -52,7 +52,8 @@ export class TilemapGeometry extends Geometry
 	        .addAttribute('aTextureCoord', buf, 0, false, 0, this.stride, 2 * 4)
 	        .addAttribute('aFrame', buf, 0, false, 0, this.stride, 4 * 4)
 	        .addAttribute('aAnim', buf, 0, false, 0, this.stride, 8 * 4)
-	        .addAttribute('aTextureId', buf, 0, false, 0, this.stride, 10 * 4);
+	        .addAttribute('aTextureId', buf, 0, false, 0, this.stride, 10 * 4)
+            .addAttribute('aAlpha', buf, 0, false, 0, this.stride, 11 * 4);
```
Comment (SukantPal) [inline src/TilemapShader.ts:56]: Can you fix this indent?

File: src/tilemap.frag [hunk 2]
```diff
@@ -12,5 +13,5 @@ void main(void)
 
    vec4 color;
    %forloop%
-   gl_FragColor = color;
-}
\ No newline at end of file
+   gl_FragColor = (color * vAlpha) + (gl_FragColor * (1.0 - vAlpha));
```
Comment (SukantPal) [inline src/tilemap.frag:16]: I don't think you're supposed to read from `gl_FragColor` if haven't set it. Can you test this:

```suggestion
   gl_FragColor = color * vAlpha;
```

File: src/Tilemap.ts [hunk 3]
```diff
@@ -10,7 +10,23 @@ import type { CanvasRenderer } from '@pixi/canvas-renderer';
 import type { IDestroyOptions } from '@pixi/display';
 import type { TilemapGeometry } from './TilemapShader';
 
-export const POINT_STRUCT_SIZE = 12;
+enum PointStruct {
```
Comment (SukantPal) [inline src/Tilemap.ts:13]: PixiJS enums are UPPER_CASE_ENUM.UPPER_CASE_PROP. Can we make this consistent?

**Review by SukantPal — CHANGES_REQUESTED** [review]
Few small pointers, great PR. Happy to approve once you follow up!

**Comment by blurymind** [comment]
Thanks for reviewing this :) I will get it updated after work

File: src/TilemapShader.ts (hunk 1 above)
Comment (blurymind) [reply src/TilemapShader.ts:56]: hmm, I dont think this is indented in the file. If you click to edit the file, the indent disappears

**Review by SukantPal — APPROVED** [review]

**Comment by SukantPal** [comment]
I'll make a release this evening.

**Comment by blurymind** [comment]
Thank you for reviewing this :tada:

**Comment by SukantPal** [comment]
Published 3.2.0, sorry for taking so long :(

**Comment by SukantPal** [comment]
The above comment was for #116 oops.

**Pull request merged**